src/rep.c

```diff
@@ -0,0 +1,210 @@
+#include "vctrs.h"
+#include "utils.h"
+
+// Initialised at load time
+static struct vctrs_arg args_times_;
+static struct vctrs_arg args_x_;
+static struct vctrs_arg* const args_times = &args_times_;
+static struct vctrs_arg* const args_x = &args_x_;
+
+static inline void stop_rep_size_oob(long long size);
+static inline void stop_rep_times_size();
+static inline void stop_rep_times_negative();
+static inline void stop_rep_times_missing();
+static inline void stop_rep_each_times_negative(R_len_t i);
+static inline void stop_rep_each_times_missing(R_len_t i);
+
+// -----------------------------------------------------------------------------
+
+static SEXP vec_rep(SEXP x, R_len_t times) {
+  if (times < 0) {
+    if (times == NA_INTEGER) {
+      stop_rep_times_missing();
+    } else {
+      stop_rep_times_negative();
+    }
+  }
+
+  const R_len_t x_size = vec_size(x);
+
+  const long long temp_size = (long long) x_size * times;
+  if (temp_size > INT_MAX) {
+    stop_rep_size_oob(temp_size);
+  }
+
+  const R_len_t size = (R_len_t) temp_size;
+
+  SEXP subscript = PROTECT(Rf_allocVector(INTSXP, size));
+  int* p_subscript = INTEGER(subscript);
+
+  R_len_t k = 0;
+
+  for (R_len_t i = 0; i < times; ++i) {
+    for (R_len_t j = 1; j <= x_size; ++j, ++k) {
+      p_subscript[k] = j;
+    }
+  }
+
+  SEXP out = vec_slice_impl(x, subscript);
+
+  UNPROTECT(1);
+  return out;
+}
+
+// [[ register() ]]
+SEXP vctrs_rep(SEXP x, SEXP times) {
+  times = PROTECT(vec_cast(times, vctrs_shared_empty_int, args_times, args_empty));
+
+  if (vec_size(times) != 1) {
+    stop_rep_times_size();
+  }
+
+  const R_len_t times_ = r_int_get(times, 0);
+
+  SEXP out = vec_rep(x, times_);
+
+  UNPROTECT(1);
+  return out;
+}
+
+// -----------------------------------------------------------------------------
+
+static SEXP vec_rep_each_uniform(SEXP x, R_len_t times) {
+  if (times < 0) {
+    if (times == NA_INTEGER) {
+      stop_rep_each_times_missing(1);
+    } else {
+      stop_rep_each_times_negative(1);
+    }
+  }
+
+  const R_len_t x_size = vec_size(x);
+
+  const long long temp_size = (long long) x_size * times;
+  if (temp_size > INT_MAX) {
+    stop_rep_size_oob(temp_size);
+  }
+
+  const R_len_t size = (R_len_t) temp_size;
+
+  SEXP subscript = PROTECT(Rf_allocVector(INTSXP, size));
+  int* p_subscript = INTEGER(subscript);
```

Comment:
`p` is normally a suffix.

Reply:
Oh really? I have been almost 100% consistent in using a prefix of `p_` throughout the vctrs code I've written

Reply:
ah I haven't noticed. I see that Google also uses a prefix for pointers.

I use a suffix for consistency with other derived variables:

```c
SEXP foo = get();

R_len_t foo_size = vec_size(foo);
int* foo_p = INTEGER(foo);

SEXP foo_names = r_names(foo);
SEXP* foo_names_p = STRING_PTR(foo_names);
```

On the other hand the prefix makes the pointer variables stand out:

```c
SEXP foo = get();

R_len_t foo_size = vec_size(foo);
int* p_foo = INTEGER(foo);

SEXP foo_names = r_names(foo);
SEXP* p_foo_names = STRING_PTR(foo_names);
```

Not sure what's best.

Reply:
I think we should use a prefix. That's what I did in readr, RPostgres, etc.